### PR TITLE
Add telemetry values to auth request and logout URIs, along with a sw…

### PR DESCRIPTION
**Pull request by polita**
…itch to turn telemetry off

### CODE REVIEW

**Comment by brentschmaltz** [comment]
There is some feedback from the community that OFF by default seems to be preferred.
https://github.com/aspnet/Security/issues/1035 (I accidentally clicked +1 on the original issue opened by pinpointtownes)

File: test/Microsoft.IdentityModel.Protocols.OpenIdConnect.Tests/OpenIdConnectMessageTests.cs [hunk 1]
```diff
@@ -165,15 +165,14 @@ public void GetSets()
             OpenIdConnectMessage message = new OpenIdConnectMessage();
             Type type = typeof(OpenIdConnectMessage);
             PropertyInfo[] properties = type.GetProperties();
-            if (properties.Length != 47)
+            if (properties.Length != 48)
                 Assert.True(true, "Number of public fields has changed from 47 to: " + properties.Length + ", adjust tests");
```
Comment (Eilon) [inline test/Microsoft.IdentityModel.Protocols.OpenIdConnect.Tests/OpenIdConnectMessageTests.cs]: Change msg to 48 as well 😄

**Comment by PinpointTownes** [comment]
> I accidentally clicked +1 on the original issue opened by pinpointtownes

Are you sure it was accidental? :trollface:

**Comment by Eilon** [comment]
@brentschmaltz BTW you can un-thumbs-up by clicking the 👍 icon at the bottom of the issue again.

File: test/Microsoft.IdentityModel.Protocols.OpenIdConnect.Tests/OpenIdConnectMessageTests.cs (hunk 1 above)
Comment (polita) [inline test/Microsoft.IdentityModel.Protocols.OpenIdConnect.Tests/OpenIdConnectMessageTests.cs]: Thank you. Fixed.

**Comment by brentschmaltz** [comment]
@PinpointTownes :-) always a pleasure.

**Comment by polita** [comment]
@brentschmaltz Can you CR?

File: src/Microsoft.IdentityModel.Protocols.OpenIdConnect/OpenIdConnectMessage.cs [hunk 2]
```diff
@@ -280,6 +304,20 @@ public string DomainHint
         }
 
         /// <summary>
+        /// Gets or sets whether parameters for the library and version are sent on the query string for this <see cref="OpenIdConnectMessage"/> instance. 
+        /// This value is set to the value of EnableTelemetryParametersByDefault at message creation time.
+        /// </summary>
+        public bool EnableTelemetryParameters
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Protocols.OpenIdConnect/OpenIdConnectMessage.cs]: The only thing I can think of is if we did

{get;set;} = EnableTelemetryParametersByDefault;

would remove the need for a few sets in some methods.

**Comment by PinpointTownes** [comment]
I don't wanna be evil, but note that thanks to the "lowest applicable version" policy used by NuGet, folks using the OIDC middleware won't get telemetry until a new OIDC package is released, unless they explicitly specify the latest IdentityModel version in their project.json :smile: 

... so maybe a static property is not necessary...? :trollface:

**Comment by polita** [comment]
@PinpointTownes Yes, it's true that you'd have to explicitly download the new package. But since people might be pulling down a new package for other reasons than the awesome new telemetry feature, giving them a single line of code way to turn it off is still desirable. 

At least, I thought for sure you'd be a proponent of that... :)

**Comment by polita** [comment]
@brentschmaltz Thanks- updated the CR and squashed.

File: src/Microsoft.IdentityModel.Protocols.OpenIdConnect/OpenIdConnectMessage.cs (hunk 2 above)
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Protocols.OpenIdConnect/OpenIdConnectMessage.cs]: @polita I don't see this in current PR. I was thinking this should be on the Property EnableTelemetryParameters

**Comment by polita** [comment]
@brentschmaltz Updated with the change to use auto-property initializers.

File: src/Microsoft.IdentityModel.Protocols.OpenIdConnect/OpenIdConnectMessage.cs [hunk 3]
```diff
@@ -280,6 +298,17 @@ public string DomainHint
...
         /// <summary>
+        /// Gets or sets whether parameters for the library and version are sent on the query string for this <see cref="OpenIdConnectMessage"/> instance. 
+        /// This value is set to the value of EnableTelemetryParametersByDefault at message creation time.
+        /// </summary>
+        public bool EnableTelemetryParameters { get; set; } = EnableTelemetryParametersByDefault;
+
+        /// <summary>
+        /// Gets or sets whether parameters for the library and version are sent on the query string for all instances of <see cref="OpenIdConnectMessage"/>. 
+        /// </summary>
+        public static bool EnableTelemetryParametersByDefault = true;
```
Comment (PinpointTownes) [inline src/Microsoft.IdentityModel.Protocols.OpenIdConnect/OpenIdConnectMessage.cs]: Out of curiosity, why not a property? :open_mouth:

Comment (polita) [inline src/Microsoft.IdentityModel.Protocols.OpenIdConnect/OpenIdConnectMessage.cs]: What would a property getter/setter get you that this doesn't?

Comment (PinpointTownes) [inline src/Microsoft.IdentityModel.Protocols.OpenIdConnect/OpenIdConnectMessage.cs]: The immense satisfaction of respecting the best practices? :trollface: 

Exposing public fields is generally discouraged (and I'm pretty sure there's even an FxCop rule for that)

Comment (Tratcher) [inline src/Microsoft.IdentityModel.Protocols.OpenIdConnect/OpenIdConnectMessage.cs]: Property implementations can be changed later without breaking callers. Fields can be changed to properties later but it requires recompiling the caller.

Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Protocols.OpenIdConnect/OpenIdConnectMessage.cs]: @Tratcher are you thinking something like, we set the property from an external source (config, etc)? Where the property could have some smarts?

Comment (polita) [inline src/Microsoft.IdentityModel.Protocols.OpenIdConnect/OpenIdConnectMessage.cs]: Updated to make it a property. Thanks for the feedback.

**Comment by brentschmaltz** [comment]
Well that was easy :-)
:shipit:

**Pull request merged**